common/src/main/java/com/wynntils/screens/maps/CustomWaypointIconScreen.java

```diff
@@ -0,0 +1,338 @@
+/*
+ * Copyright © Wynntils 2024.
+ * This file is released under LGPLv3. See LICENSE for full license details.
+ */
+package com.wynntils.screens.maps;
+
+import com.mojang.blaze3d.vertex.PoseStack;
+import com.wynntils.core.WynntilsMod;
+import com.wynntils.core.components.Services;
+import com.wynntils.core.text.StyledText;
+import com.wynntils.screens.base.WynntilsGridLayoutScreen;
+import com.wynntils.screens.base.widgets.TextInputBoxWidget;
+import com.wynntils.screens.maps.widgets.CustomIconWidget;
+import com.wynntils.services.mapdata.providers.json.JsonIcon;
+import com.wynntils.utils.MathUtils;
+import com.wynntils.utils.colors.CommonColors;
+import com.wynntils.utils.mc.McUtils;
+import com.wynntils.utils.render.FontRenderer;
+import com.wynntils.utils.render.RenderUtils;
+import com.wynntils.utils.render.type.HorizontalAlignment;
+import com.wynntils.utils.render.type.TextShadow;
+import com.wynntils.utils.render.type.VerticalAlignment;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.List;
+import java.util.regex.Pattern;
+import net.minecraft.client.gui.GuiGraphics;
+import net.minecraft.client.gui.components.Button;
+import net.minecraft.client.gui.components.Tooltip;
+import net.minecraft.client.gui.screens.Screen;
+import net.minecraft.client.resources.language.I18n;
+import net.minecraft.network.chat.Component;
+
+public class CustomWaypointIconScreen extends WynntilsGridLayoutScreen {
+    private static final int ICONS_PER_PAGE = 16;
+    private static final int SCROLLBAR_HEIGHT = 40;
+    private static final float SCROLL_FACTOR = 10f;
+    private static final Pattern ICON_NAME_PATTERN = Pattern.compile("^[a-z0-9/._-]+$");
```

Comment:
```suggestion
    // Accepted resource location name pattern `^[a-z0-9/._-]+`
    // This is made stricter to avoid weird names
    private static final Pattern ICON_NAME_PATTERN = Pattern.compile("^[a-z0-9-]+$");
```